### DISCUSSION TITLE
Unsure how to initialize a value using IOE before running reinterpret in my effect handler function.

### DISCUSSION THREAD

**Discussion by rmullin7286** [haskell-effectful/effectful#229] [Q&A]
Basically I have an effect of the following form:

```
data ApiKey = ApiKey { accesskey :: String, secretKey :: String }

data Request

data Reponse

data Api :: Effect where
    CallEndpoint :: Request -> Api m Response
```

Before the effect can be run, I need to call /login with the ApiKey to initialize the session state. Something like

```
runApiIO :: (IOE :> es) => Eff (Api : es) a -> Eff es a
runApiIO = do
    session <- loginWithCredentials
    reinterpret (evalState session) $ \env -> \case
        CallEndpoint r -> do
            session <- get
            resp <- post session r
            return $ responseBodyToJson resp
```

This doesn't seem to work. I notice looking at the source for things like the State effect that they use `unsafeEff_` to initialize mvars and other variables that require IO. I'd rather not have to lean on unsafe calls if I don't have to.

Is the recommended way to do this to create a function `loginWithCredentials :: IOE :> es => ApiKey -> Eff es Session`, call that beforehand, and then change runAPIIO to be `IOE :> es => Session -> Eff (Api : es) a -> Eff es a`?

**Comment by arybczak** [comment]
> This doesn't seem to work

What exactly do you mean by that? `runApiIO` looks fine to me (well, what you wrote won't compile since you need to pass the argument of `runApiIO` to `reinterpret`, but the idea looks good).

**Comment by ocharles** [comment]
When I do things like this, I tend to prefer:

```haskell
runApiIO :: (IOE :> es) => Eff (Api : es) a -> Eff es a
runApiIO = do
    reinterpret withSession $ \env -> \case
        CallEndpoint r -> do
            session <- get
            resp <- post session r
            return $ responseBodyToJson resp
  where
    withSession m = do
      session <- loginWithCredentials
      evalState session m
```

I think your problem is just that you forgot to bind the argument in `runApiIO`:

```haskell
runApiIO :: (IOE :> es) => Eff (Api : es) a -> Eff es a
runApiIO m = do
    session <- loginWithCredentials
    flip (reinterpret (evalState session)) m $ \env -> \case
        CallEndpoint r -> do
            session <- get
            resp <- post session r
            return $ responseBodyToJson resp
```

Something like that
